File: my-folder/0098-validate-binary-search-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        
        return helper(root, null, null);
    }
    
    public boolean helper(TreeNode root, Integer upper, Integer lower){
        if(root==null){
            return true;
        }
        
         if(upper!=null&&root.val>=upper||lower!=null&&root.val<=lower){
            return false;
        }
        
        return helper(root.left, root.val, lower) && helper(root.right, upper, root.val);
    }
}

File: my-folder/0098-validate-binary-search-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.isBST(root, None, None)
    def isBST(self, root, minNode,maxNode):
        if root is None:
            return True
        if minNode is not None and root.val <= minNode.val:
            return False
        if maxNode is not None and root.val >= maxNode.val:
            return False
        return self.isBST(root.left, minNode, root) and self.isBST(root.right, root, maxNode)

